keps/0030-immutable-parameters.md

```diff
@@ -0,0 +1,173 @@
+
+### Goals
+
+Make it possible for an operator developer to specify that a parameter can not be changed after the installation
+
+### Non-Goals
+
+- Allow modification of immutable parameters after the initial deploy plan is finished
+- Constants that can be used in templating but can not be changed by the user. Immutable parameters can always be set by the user on installation.
+- Immutable Parameters that can be set after the installation - There is a use case for parameters that have an undefined state until they are first used, for example in a specific plan, but can not be changed afterwards. This KEP requires a parameter to receive a value at installation time, either by using a default or provided by the user.
+
+## Proposal
+
+An `immutable` property is added to the parameter definition. If that property is true, the parameter needs to either have a default value or has to be explicitly provided by the user when the operator is installed. The parameter can not be updated with `kudo update -p` anymore.
+
+```yaml
+parameters:
+  - name: NUM_TOKENS
+    description: The number of tokens on each node of the cluster
+    default: 256
+    immutable: true
+```
+
+The default value for `immutable` is `false`.
+
+### User Stories
+
+- [#652](https://github.com/kudobuilder/kudo/issues/652) Make it possible to specify immutable parameter
+
+- NUM_TOKENS in the Cassandra operator
+- STORAGE_CLASS in the Zookeeper operator
+- DISK_SIZE, STORAGE_CLASS and PERSISTENT_STORAGE in the Kafka operator
+
+### Implementation Details/Notes/Constraints
+
+#### Installation
+- If a parameter definition is immutable, it must either have a default value or needs to be marked as required.
+- If no value for an immutable parameter is given on installation, KUDO copies the default value from the OperatorVersion into the instance. This makes sure that the parameter value will never change, even if instance is changed to use a newer OperatorVersion that has different defaults.
+
+#### Updates
+- KUDO CLI will not perform any checks with regards to immutability on installation or updates
+- The admission webhook verifies if a parameter is immutable
+  - If the value of any immutable parameter changes from the current value it rejects the whole update
+  - If the value of all immutable parameters are the same as their current values it allows the update
+
+#### Upgrades
+- Upgrades to a new OperatorVersions:
+  - A parameter can be made immutable in a newly released OperatorVersion.
+    - When executing `kudo upgrade` the user has to explicitly set the value for parameters that are changed from mutable to immutable
+    - If an parameter is changed from mutable to immutable and the user does not explicitly specifies a parameter the upgrade will abort with an error message
```

Comment:
```suggestion
    - If an parameter is changed from mutable to immutable and the user does not explicitly specify a parameter the upgrade will abort with an error message
```
fixed the grammar... however I'm not sure I like this...

Reply:
after some thought... if you want user input.. perhaps an ack flag of some sort. `--force` or `--params-upgrade`

specifying every param is the cognitive load we should avoid

Reply:
Mmmm, i'm torn on this one. I really want to make sure users do not accidentally set an immutable parameter they don't know about.

I see two options:
1. The user has to specify all new immutable parameters explicitly. This has to be done for immutable parameters without a default in any case, and I don't think there will be that many new immutable parameters introduced to a new operator version.
2. Add a new parameter to `kudo upgrade`. If this is specified, the upgrade process would automatically use the default values. I think this would be an extra parameter for a very specific use case that won't be used often, and it may lead to users adding this parameter to their tooling, ignoring new immutable parameters.

As said above, I think the use case here will happen very seldom and I think it's acceptable for users to specify all parameters. Especially if parameter-files are used. KUDO should error with all missing parameters so this would only be one round trip: Call `kudo upgrade ....` -> See missing parameters -> Add parameters to param-file -> Call `kudo upgrade` again.

If you insist on a parameter I'm ok with it but I would prefer to not use one.

Reply:
I don't know how it is possible to know the freq or likelihood of this situation.   we also don't know the magnitude of impact (number of changes for an upgrade).  It seems that the goal is to make sure the user is aware of this situation.   It is a reasonable assumption that 1 of 2 things will happen for an upgrade of this kind... either they know or they don't.  IF they don't they will try to do an upgrade and an error will occur... now they are aware.  I suggest that the error returned is a list of all violations (all params that were mutable that became immutable).  Now they are aware.. to force them to specify each one is to punish the user... (and to punish them for something the operator dev did.. not them).  A flag that indicates they are aware should be good enough.   I could imagine that if the error was generic that in your model, specifying each param is ensuring they know of each and every one... but it would like result in a try/fail/trywith(a)/fail/trywith(a,b)/fail/trywith(a,b,c)/succeed.   We should provide an error with a list of all parameters to be aware of... if they have that information, then why punish them with jump through hoops?
your model seems to expect param-file... I'm not sure if they will have this... also, they may be upgrading without any changed params (so no param-file at all).   
I think the better UX is to have the ack flag.
While I think it is a much better UX... It isn't a blocker for me.